### DISCUSSION TITLE
How can I focus the Listing View and Decompiler View programatically?

### DISCUSSION THREAD

**Discussion by jc874** [NationalSecurityAgency/ghidra#2929] [Q&A]
Hi,

I would like to send focus to the Listing View or Decompiler view from my Java plugin but don't want to use global keybindings via the `Edit->Tool options...->Key bindings` window.

Essentially I'd like to be able to jump from my plugin, the listing view and decompiler view using only key bindings set in my plugin (**not** globally set via the `Tool options...` window). The reason why I don't want to set global key bindings is because I'd like my plugin to be fully standalone and not interfere with users key bindings. So for example if my plugin has focus and it receives the key `d` then I'd like to switch focus to the Decompiler window.

How can this be achieved?

**Comment by dragonmacher** [comment]
This should be possible a couple different ways.  The typical way is to create a DockingAction that is local to you provider (you register local actions via `Tool.addLocalAction(ComponentProvider provider, DockingActionIf action)`.   This action will only be enabled for you provider.   You can assign a keybinding to your action as desired.    You could also do this with custom `KeyListener`s on your view components.    We generally do not use this second approach, as there is no way for users to change this behavior from the tool.

It is not clear why you need the ability to jump to these other widgets from your widget.   Each provider can have a keybinding assigned by the user that allows it to be focused from anywhere.  This allows users to set the keybindings they prefer.   Having another means of doing this may be confusing to the user.

**Reply by jc874** [reply]
Thank you for your reply. With the `addLocalAction()` method I can see how I can create my own custom `Action` and assign it to a keybinding but how would I make that custom `Action` actually switch/focus to something like the Decompiler view?

As for why I want to implement this functionality, it's mainly for my very niche use 😃

**Comment by dragonmacher** [comment]
You should be able to use `public ComponentProvider getComponentProvider(String name);` on the tool to get the provider by name.   Once you have the provider, you can then call `public void toFront(ComponentProvider componentProvider);`.

**Reply by jc874** [reply]
hmmm this feels close. I have the following few lines of code in my plugin now...

`ComponentProvider decompiler = pluginTool.getComponentProvider("Decompiler");
pluginTool.toFront(decompiler);`

... but the decompiler window isn't taking the focus after this code is ran. `decompiler` looks like it has a valid value (not null)

**Reply by jc874** [reply]
Ah, its working now if I use `decompiler.requestFocus()` just after the `toFront()` call!
